### PR TITLE
WIP: Bugfixes in figure parsing for chord symbols

### PR DESCRIPTION
**Pull request by a-papadopoulos**
I fixed a couple of bugs I encountered while parsing a large dataset of lead sheets.

Basically, I want to ensure that given a ChordSymbol ch, music21 can always create a ChordSymbol given the figure ch.figure, and that the resulting chord symbol is always equivalent to the first.

I have included tests for the several cases I encountered, without any reference to the original dataset.

### CODE REVIEW

**Comment by a-papadopoulos** [comment]
Added two more bugfixes!

File: music21/abcFormat/translate.py [hunk 1]
```diff
@@ -1031,20 +1031,6 @@ def testCleanFlat(self):
         self.assertEqual(cs.bass(), pitch.Pitch('B-2'))
         self.assertEqual(cs.pitches[0], pitch.Pitch('B-2'))
 
-        # common.cleanedFlatNotation() shouldn't be called by
```
Comment (mscuthbert) [inline music21/abcFormat/translate.py]: Can you explain why these tests are removed?

File: music21/chord/__init__.py [hunk 2]
```diff
@@ -4770,6 +4770,26 @@ def volume(self, expr):
 
     # --------------------------------------------------------------------------
 
+def sort_pitches(notes_or_pitches, key_pitch):
```
Comment (mscuthbert) [inline music21/chord/__init__.py]: naming convention is camel case.  Can you explain why this function is necessary?

File: music21/chord/__init__.py [hunk 3]
```diff
@@ -3698,7 +3698,7 @@ def sortDiatonicAscending(self, *, inPlace=False):
             returnObj = self
         else:
             returnObj = copy.deepcopy(self)
-        returnObj._notes.sort(key=lambda x: (x.pitch.diatonicNoteNum, x.pitch.ps))
+        sortDiatonicAscending(returnObj._notes)
```
Comment (mscuthbert) [inline music21/chord/__init__.py]: This request takes a documented and tested method and splits it into two functions, one of which has no documentation or tests.

File: music21/harmony.py [hunk 4]
```diff
@@ -1458,8 +1474,7 @@ class ChordSymbol(Harmony):
     (<music21.pitch.Pitch A-2>, <music21.pitch.Pitch C-3>, <music21.pitch.Pitch E-3>)
 
     >>> harmony.ChordSymbol('F-dim7').pitches
-    (<music21.pitch.Pitch F-2>, <music21.pitch.Pitch A--2>,
-     <music21.pitch.Pitch C--3>, <music21.pitch.Pitch E---3>)
+    (<music21.pitch.Pitch F-2>, <music21.pitch.Pitch A--2>, <music21.pitch.Pitch C--3>, <music21.pitch.Pitch E---3>)
```
Comment (mscuthbert) [inline music21/harmony.py]: fails PEP8 line length.

File: music21/harmony.py [hunk 5]
```diff
@@ -1321,6 +1325,18 @@ def removeChordSymbols(chordType):
     '''
     del CHORD_TYPES[chordType]
 
+def _get_alteration(alteration):
```
Comment (mscuthbert) [inline music21/harmony.py]: method requires documentation and tests.

File: music21/harmony.py [hunk 6]
```diff
@@ -1541,22 +1556,21 @@ def _adjustOctaves(self, pitches):
...
-            pitches[1] = pitch.Pitch(pitches[1].name + str(pitches[1].octave + 1))
-            pitches[3] = pitch.Pitch(pitches[3].name + str(pitches[3].octave + 1))
-            pitches[5] = pitch.Pitch(pitches[5].name + str(pitches[5].octave + 1))
+            pitches[1].octave += 1
+            pitches[3].octave += 1
+            pitches[5].octave += 1
         else:
             return pitches
 
-        c = chord.Chord(pitches)
```
Comment (mscuthbert) [inline music21/harmony.py]: please revert this -- I prefer one way of doing the sort -- the speed overhead of creating an extra chord is not enough to justify extra documentation and support concerns.

File: music21/harmony.py [hunk 7]
```diff
@@ -1605,6 +1619,10 @@ def _adjustPitchesForChordStepModifications(self, pitches):
 
         rootPitch = self.root()
         sc = scale.MajorScale(rootPitch)
+        # in case of inversion, the root should be updated, so make sure it's
+        # still accurate
+        if not rootPitch is pitches[0]:
+            print('hey')
```
Comment (mscuthbert) [inline music21/harmony.py]: test code left in.

File: music21/harmony.py [hunk 8]
```diff
@@ -1691,7 +1709,7 @@ def typeAlter(hD):
                 degree = degree.replace('A', '') #A is for 'Altered'
                 if hD.degree == int(degree):
                     # transpose by semitones (positive for up, negative for down)
-                    p = p.transpose(hD.interval)
+                    p.transpose(hD.interval, inPlace=True),
```
Comment (mscuthbert) [inline music21/harmony.py]: This changes the original pitch -- is this what is wanted?  Need tests to be sure.

File: music21/harmony.py [hunk 9]
```diff
@@ -1809,27 +1837,19 @@ def _parseFigure(self):
             bass = bass.replace('/', '')
             self.bass(bass)
             #remove the root and bass from the string and any additions/omitions/alterations/
-            remaining = st.replace(m2.group(), '')
+            # get the match on st, should match since st is a substring of
+            # prelimFigure
+            m3 = re.search(r'/[A-Ga-g][#-]*', st)
+            assert m3
```
Comment (mscuthbert) [inline music21/harmony.py]: bare asserts not allowed in music21 code base.  Does not tell the user what to do to fix.  Raise an error with a message that can be trapped.

File: music21/harmony.py [hunk 10]
```diff
@@ -1809,27 +1837,19 @@ def _parseFigure(self):
...
-        if 'omit' in remaining or 'subtract' in remaining:
-            degree = remaining[remaining.index('omit') + 4: ]
-            self.addChordStepModification(ChordStepModification('subtract', int(degree)))
-            return
 
         st = st.replace(',', '')
-
-        if 'b' in st or '#' in st:
+        if 'add' in st or 'alter'in st or 'omit' in st or 'subtract' in st:
+            splitter = re.compile('((?:alter|add|omit|subtract)[b#]*[1-9]+)')
```
Comment (mscuthbert) [inline music21/harmony.py]: compilation should not take place in a loop, or it loses its utility.

File: music21/harmony.py [hunk 11]
```diff
@@ -1711,23 +1729,33 @@ def typeAlter(hD):
...
-            elif hD.modType == 'subtract':
+            elif hD.modType == 'subtract' or hD.modType == 'omit':
                 typeSubtract(hD)
             elif hD.modType == 'alter':
                 typeAlter(hD)
 
         return tuple(pitches)
 
     def _getKindFromShortHand(self, sH):
+        all_abbr = [a for type in CHORD_TYPES.values() for a in type[1]]
```
Comment (mscuthbert) [inline music21/harmony.py]: this is a substantial enough change it will need a battery of tests to make sure it works.  Thanks!

Comment (mscuthbert) [reply music21/harmony.py]: (and documentation -- it should have had it originally; that was an oversight)

File: music21/harmony.py [hunk 12]
```diff
@@ -1944,13 +1960,15 @@ def _updatePitches(self):
...
         if self._notationString():
-            pitches = fbScale.getSamplePitches(self._overrides['root'], self._notationString())
+            new_pitches = fbScale.getSamplePitches(self._overrides['root'],
+                                           self._notationString())
             # remove duplicated bass note due to figured bass method.
-            pitches.pop(0)
-        else:
-            pitches = []
-            pitches.append(self._overrides['root'])
+            pitches.extend(new_pitches[2:])
```
Comment (mscuthbert) [inline music21/harmony.py]: what if new_pitches is length 1?  Is this possible?

File: music21/harmony.py [hunk 13]
```diff
@@ -2474,6 +2495,315 @@ def chordSymbolSetsBassOctave(self):
...
+            </degree>
+          </harmony>
+        """
+
+        MP = musicxml.xmlToM21.MeasureParser()
+        mxHarmony = EL(xmlString)
+
+        cs1 = MP.xmlToChordSymbol(mxHarmony)
+        cs2 = ChordSymbol(cs1.figure)
+        cs3 = ChordSymbol('G7 subtract 5 add b9 add #9 add #11 add b13')
```
Comment (mscuthbert) [inline music21/harmony.py]: what if spaces are omitted between "add #9" etc.?  This is common.

File: music21/harmony.py [hunk 14]
```diff
@@ -2474,6 +2495,315 @@ def chordSymbolSetsBassOctave(self):
...
+        cs1 = MP.xmlToChordSymbol(mxHarmony)
+        cs2 = ChordSymbol(cs1.figure)
+        cs3 = ChordSymbol('C/E add 2')
+
+        self.assertEqual(pitches, cs1.pitches)
+        self.assertEqual(pitches, cs2.pitches)
+        self.assertEqual(pitches, cs3.pitches)
+
+    def testNinth(self):
+        """
```
Comment (mscuthbert) [inline music21/harmony.py]: excellent!

File: music21/musicxml/m21ToXml.py [hunk 15]
```diff
@@ -4576,7 +4575,7 @@ def chordSymbolToXml(self, cs):
           <kind>suspended-fourth</kind>
           <degree>
             <degree-value>9</degree-value>
-            <degree-alter />
+            <degree-alter>0</degree-alter>
```
Comment (mscuthbert) [inline music21/musicxml/m21ToXml.py]: degree-alter can be empty, or at least I have seen it empty in the wild -- does that still work?

**Comment by a-papadopoulos** [comment]
Thanks for the extensive reviewing! I'm actually fixing some extra bugs, so I'll reply to everything once I'm finished with them too.

File: music21/harmony.py (hunk 9 above)
Comment (a-papadopoulos) [reply music21/harmony.py]: I want to comment on this: why are asserts not allowed? I agree it is completely wrong to use asserts to convey error messages to users. They are useful however for debugging purposes, in order to formally specify invariants that should always be satisfied in any execution, i.e. regardless of user "mistakes". In that sense, they serve the same purpose as unit tests. In fact, running a piece of code and verifying that all asserts pass is a valuable unit test in itself.

Here, I'm just saying that 'm3' should always match since 'st' is a substring of 'prelimFigure'. I could just write it as a comment, but I might as well check it. If this assert fails, it's a programming error of mine, and not a user error, so this should never be reported to the user, but rather, the code should be sufficiently tested to ensure that the assert never fails.

Anyway, this might boil down to coding style, so I'm happy to simply remove this line. Furthermore, Python has the nasty behaviour of enabling asserts by default (i.e. having '__debug__' on by default), when the opposite should be the default, so that the asserts are only verified at testing phase, and this invalidates my argument to some extent.

File: music21/harmony.py (hunk 8 above)
Comment (a-papadopoulos) [reply music21/harmony.py]: Previous version was a no-op, the resulting pitch was ignored and the chord modification not applied. This was caught by one of the tests I added. I can certainly rewrite it in a functional, non-inPlace mode (which seems to be in the spirit of this method return the new list of pitches).

File: music21/abcFormat/translate.py (hunk 1 above)
Comment (a-papadopoulos) [reply music21/abcFormat/translate.py]: These tests where testing the fact that the root with 'bb3' should not be interpreted as b-3. Since I fixed some bugs with chord alterations, this interfered with the new code, so I lazily removed them.
Shouldn't have, so I put them back and corrected it: now b3 is interpreted as a chord alteration, and D# becomes D. Also, the pitches are one octave lower because of the '_hasPitchAboveC4' test.

**Comment by mscuthbert** [comment]
Thanks -- I'll have a review by Saturday!

**Comment by a-papadopoulos** [comment]
I'm just discovering that I've lost many edits, for a reason I cannot fathom! So this is not complete, and quite likely is missing some bugfixes and tests. Please do not merge until I've fixed that! Sorry :(

**Comment by a-papadopoulos** [comment]
All right, I think I've recovered this! Hopefully things are ok now!

File: music21/harmony.py (hunk 6 above)
Comment (mscuthbert) [reply music21/harmony.py]: What I mean is please remove the separate function `sortDiatonicAscending` outside of the Chord object -- which doesn't return anything and only works on lists not on tuples, etc. I much prefer the existing way of saying "hey chord, sort your pitches diatonically ascending" than having it as a separate function within the `chord` module

File: music21/harmony.py (hunk 10 above)
Comment (mscuthbert) [reply music21/harmony.py]: compilation should still be a private CONSTANT at the class level.

File: music21/musicxml/m21ToXml.py (hunk 15 above)
Comment (mscuthbert) [reply music21/musicxml/m21ToXml.py]: thanks!

File: music21/chord/__init__.py (hunk 3 above)
Comment (mscuthbert) [reply music21/chord/__init__.py]: See below -- I still prefer this to be in the chord itself.

File: music21/harmony.py [hunk 16]
```diff
@@ -1321,6 +1326,47 @@ def removeChordSymbols(chordType):
...
+    ('add', 9, -1)
+
+    >>> harmony._getAlteration('alter#5')
+    ('alter', 5, 1)
+
+    >>> harmony._getAlteration('alter##5')
+    ('alter', 5, 2)
+
+    """
+    if alteration != '':
```
Comment (mscuthbert) [inline music21/harmony.py]: short circuit the loop here.  `if alteration == '': return None`

Comment (mscuthbert) [reply music21/harmony.py]: Or better: `if not alteration: return None`

File: music21/harmony.py [hunk 17]
```diff
@@ -1458,8 +1504,8 @@ class ChordSymbol(Harmony):
     (<music21.pitch.Pitch A-2>, <music21.pitch.Pitch C-3>, <music21.pitch.Pitch E-3>)
 
     >>> harmony.ChordSymbol('F-dim7').pitches
-    (<music21.pitch.Pitch F-2>, <music21.pitch.Pitch A--2>,
-     <music21.pitch.Pitch C--3>, <music21.pitch.Pitch E---3>)
+    (<music21.pitch.Pitch F-2>, <music21.pitch.Pitch A--2>, \
+<music21.pitch.Pitch C--3>, <music21.pitch.Pitch E---3>)
```
Comment (mscuthbert) [inline music21/harmony.py]: Why this change?  the documentation will look worse with it.

File: music21/harmony.py [hunk 18]
```diff
@@ -1541,22 +1587,37 @@ def _adjustOctaves(self, pitches):
...
+        """
+        root = self._overrides['root']
+        bass = self._overrides['bass']
+        bass.octave = root.octave
 
-        return list(c.pitches)
+        if bass.ps > root.ps:
+            bass.octave -= 1
+
+        assert bass.ps < root.ps
```
Comment (mscuthbert) [inline music21/harmony.py]: no bare asserts (music21 style guide)

File: music21/harmony.py [hunk 19]
```diff
@@ -1541,22 +1587,37 @@ def _adjustOctaves(self, pitches):
...
-        c = c.sortDiatonicAscending()
+        sortDiatonicAscending(pitches)
+
+        return pitches
+
+    def _assignOctaveForBass(self):
+        """
+        Assigns the highest octave to the bass that ensure it remains lower
+        than the root.
+        """
```
Comment (mscuthbert) [inline music21/harmony.py]: needs doctest

File: music21/harmony.py [hunk 20]
```diff
@@ -1711,23 +1778,70 @@ def typeAlter(hD):
         for hD in chordStepModifications:
             if hD.modType == 'add':
                 typeAdd(hD)
-            elif hD.modType == 'subtract':
+            elif hD.modType == 'subtract' or hD.modType == 'omit':
```
Comment (mscuthbert) [inline music21/harmony.py]: `elif hD.modType in ('subtract', 'omit'):`

File: music21/harmony.py [hunk 21]
```diff
@@ -1711,23 +1778,70 @@ def typeAlter(hD):
...
                 typeSubtract(hD)
             elif hD.modType == 'alter':
                 typeAlter(hD)
 
         return tuple(pitches)
 
     def _getKindFromShortHand(self, sH):
+        """
+        Tries to identify a chord kind abbreviation in the given shorthand,
```
Comment (mscuthbert) [inline music21/harmony.py]: Thank you for the docs here where the existing code didn't meet current style specifications

File: music21/harmony.py [hunk 22]
```diff
@@ -1944,27 +2045,37 @@ def _updatePitches(self):
...
+                # weird chord here, but it shouldn't fail).
+                ignoreInversion = True
+
+            if ignoreInversion:
                 inversionNum = None
-                self._overrides['bass'].octave = 2
-                    # arbitrary octave, must be below root,
-                    # which was arbitrarily chosen as 3 above
+                # Make bass lower than the root
+                self._assignOctaveForBass()
```
Comment (mscuthbert) [inline music21/harmony.py]: I like!

File: music21/harmony.py [hunk 23]
```diff
@@ -2171,26 +2280,32 @@ def inversionIsValid(self, inversion):
...
+            or kind in ninths
+            or kind in elevenths
+            or kind in thirteenths
             ):
             return True
         elif (inversion in (2, 1)
-                and not self.chordKind == 'pedal'):
+              and not kind in ('pedal', 'power', 'suspended-second',
+                               'suspended-fourth', 'suspended-fourth-seventh')):
+            # Inversions are not quite defined for those chord types, so it's
```
Comment (mscuthbert) [inline music21/harmony.py]: super!

File: music21/harmony.py [hunk 24]
```diff
@@ -2474,6 +2589,443 @@ def chordSymbolSetsBassOctave(self):
...
+        
+        # self.assertEqual(cs1.pitches, pitches)
+        # self.assertEqual(cs2.pitches, pitches)
+
+        # self.assertEqual(cs1.root(), cs2.root())
+        # self.assertEqual(cs1.bass(), cs2.bass())
+
+
+
+    def testAddSubtractAlterations(self):
```
Comment (mscuthbert) [inline music21/harmony.py]: these tests are AWESOME.  Thank you!

File: music21/musicxml/m21ToXml.py [hunk 25]
```diff
@@ -4495,7 +4495,6 @@ def noChordToXml(self, cs):
     def chordSymbolToXml(self, cs):
         '''
         Convert a ChordSymbol object to an mxHarmony object.
-
```
Comment (mscuthbert) [inline music21/musicxml/m21ToXml.py]: restore space before doctest for documentation generation.

File: music21/harmony.py (hunk 17 above)
Comment (a-papadopoulos) [reply music21/harmony.py]: I'm not sure why I need this, it's just that the doctest fails otherwise ([as described here](https://stackoverflow.com/questions/13395610/wrapping-python-doctest-results-that-are-longer-than-80-characters) for example). Maybe it's just something with my local config, I'll remove it and see if it still passes.

Comment (mscuthbert) [reply music21/harmony.py]: Music21 doctests (when run through its system) are space agnostic -- it doesn't matter if there are extra spaces or not.

File: music21/harmony.py (hunk 16 above)
Comment (a-papadopoulos) [reply music21/harmony.py]: Done!

**Comment by a-papadopoulos** [comment]
Sorry I can't reply to the conversation about the sort function, so I'm doing this here. I still don't understand your request actually.
Creating a Chord object involves quite a bit of stuff, and I don't want to do that just to sort a list of pitches, and this actually caused some bugs (which was the reason I did that in the first place).
I can remove the separate function and keep only the method as you're suggesting. However, I won't be calling that method, but instead, will use the one-liner that does the inplace sorting. That introduces some slight code duplication though.

File: music21/harmony.py (hunk 19 above)
Comment (mscuthbert) [reply music21/harmony.py]: super!  This is wonderful!

**Comment by mscuthbert** [comment]
About the sort functions -- what I've learned from running music21 for over a decade is that every additional function/method/etc. incurs support time and costs for me over the next several years, so it's not worth it to me to speed up one function a bit by making two different functions that do similar but slightly different things.  Right now we have one documented place for sorting pitches ascending or descedning according to different criteria -- that's in the Chord object.  This can't be removed while remaining backwards compatible.  So I'd rather use the Chord object's sorting routines even if it means making an additional chord and then discarding it or even making a copy of the pitches before doing so.  Thanks

**Comment by mscuthbert** [comment]
Pinging -- this was very close.

**Comment by a-papadopoulos** [comment]
Wow, I got caught with so many other things that I eventually forgot about this. Yes, it was close, I'll try to close this in the next few days.

**Comment by mscuthbert** [comment]
Totally understand!!

**Comment by mscuthbert** [comment]
Closing and reopening to get new checks on Github Actions.

**Comment by jacobtylerwalls** [comment]
Happy New Year, @a-papadopoulos. It was reassuring to see that you hit upon some of the same solutions to various issues I was working on.

I ran your tests against the current version to see what is left to address. Here are the remaining failures:

- [ ] two failures re: support for 'add' and 'subtract' in `ChordStepModification`
- [ ] several failures re: 'add2' or '/G' or non-chord tones putting the modified note in the correct octave
- [ ] `testPower` shows a failure to add an E from 'Apower/E'
- [ ] `testRootBassParsing` now shows a different failure than name of the test implies--having to do with dominant vs. dominant-seventh aliases, perhaps?
- [ ] missing octave numbers on bass notes in `testChordWithoutKind` and `testSus2Bass`

Then there are some failures I think should pass just fine once they are altered a little:
- [ ] `testChordFlatSharpInFigure`: change A2 to A3 in `ChordSymbol(G+9)`
- [ ] `testChordWithoutKind`: change C3 to C4

<details>
ERROR: testAddSubtractAlterations (__main__.Test)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jwalls/music21/music21/harmony.py", line 3058, in testAddSubtractAlterations
    ch1 = ChordSymbol('F7 add 4 subtract 3')
  File "/Users/jwalls/music21/music21/harmony.py", line 1560, in __init__
    super().__init__(figure, **keywords)
  File "/Users/jwalls/music21/music21/harmony.py", line 208, in __init__
    self._parseFigure()
  File "/Users/jwalls/music21/music21/harmony.py", line 1856, in _parseFigure
    ChordStepModification('add', int(degree)), updatePitches=False)
ValueError: invalid literal for int() with base 10: '4subtract3'

ERROR: testChordStepFromFigure (__main__.Test)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jwalls/music21/music21/harmony.py", line 2694, in testChordStepFromFigure
    self.runTestOnChord(xmlString, figure, pitches)
  File "/Users/jwalls/music21/music21/harmony.py", line 2633, in runTestOnChord
    cs2 = ChordSymbol(cs1.figure)
  File "/Users/jwalls/music21/music21/harmony.py", line 1560, in __init__
    super().__init__(figure, **keywords)
  File "/Users/jwalls/music21/music21/harmony.py", line 208, in __init__
    self._parseFigure()
  File "/Users/jwalls/music21/music21/harmony.py", line 1856, in _parseFigure
    ChordStepModification('add', int(degree)), updatePitches=False)
ValueError: invalid literal for int() with base 10: 'b9add#9add#11addb13'

FAIL: testChordFlatSharpInFigure (__main__.Test)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jwalls/music21/music21/harmony.py", line 2794, in testChordFlatSharpInFigure
    self.runTestOnChord(xmlString, figure, pitches)
  File "/Users/jwalls/music21/music21/harmony.py", line 2637, in runTestOnChord
    self.assertEqual(pitches, cs2.pitches)
AssertionError: Tuples differ: (<mus[38 chars]itch A2>, <music21.pitch.Pitch B2>, <music21.p[37 chars] F3>) != (<mus[38 chars]itch B2>, <music21.pitch.Pitch D#3>, <music21.[37 chars] A3>)

First differing element 1:
<music21.pitch.Pitch A2>
<music21.pitch.Pitch B2>

  (<music21.pitch.Pitch G2>,
-  <music21.pitch.Pitch A2>,
   <music21.pitch.Pitch B2>,
   <music21.pitch.Pitch D#3>,
-  <music21.pitch.Pitch F3>)
?                          ^

+  <music21.pitch.Pitch F3>,
?                          ^

+  <music21.pitch.Pitch A3>)

FAIL: testChordStepBass (__main__.Test)
This tests a bug where the chord modification (add 2) was placed at a
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jwalls/music21/music21/harmony.py", line 2855, in testChordStepBass
    self.runTestOnChord(xmlString, figure, pitches)
  File "/Users/jwalls/music21/music21/harmony.py", line 2636, in runTestOnChord
    self.assertEqual(pitches, cs1.pitches)
AssertionError: Tuples differ: (<music21.pitch.Pitch E3>, <music21.pitch.Pitch G3>, <music21.p[36 chars] D4>) != (<music21.pitch.Pitch D2>, <music21.pitch.Pitch E3>, <music21.p[36 chars] C4>)

First differing element 0:
<music21.pitch.Pitch E3>
<music21.pitch.Pitch D2>

+ (<music21.pitch.Pitch D2>,
- (<music21.pitch.Pitch E3>,
? ^

+  <music21.pitch.Pitch E3>,
? ^

   <music21.pitch.Pitch G3>,
-  <music21.pitch.Pitch C4>,
?                          ^

+  <music21.pitch.Pitch C4>)
?                          ^

-  <music21.pitch.Pitch D4>)

FAIL: testChordWithBass (__main__.Test)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jwalls/music21/music21/harmony.py", line 2779, in testChordWithBass
    self.runTestOnChord(xmlString, figure, pitches)
  File "/Users/jwalls/music21/music21/harmony.py", line 2636, in runTestOnChord
    self.assertEqual(pitches, cs1.pitches)
AssertionError: Tuples differ: (<mus[39 chars]tch A2>, <music21.pitch.Pitch C#3>, <music21.pitch.Pitch E3>) != (<mus[39 chars]tch A3>, <music21.pitch.Pitch C#4>, <music21.p[36 chars] G4>)

First differing element 1:
<music21.pitch.Pitch A2>
<music21.pitch.Pitch A3>

Second tuple contains 1 additional elements.
First extra element 4:
<music21.pitch.Pitch G4>

  (<music21.pitch.Pitch G2>,
-  <music21.pitch.Pitch A2>,
?                        ^

+  <music21.pitch.Pitch A3>,
?                        ^

-  <music21.pitch.Pitch C#3>,
?                         ^

+  <music21.pitch.Pitch C#4>,
?                         ^

-  <music21.pitch.Pitch E3>)
?                        ^ ^

+  <music21.pitch.Pitch E4>,
?                        ^ ^

+  <music21.pitch.Pitch G4>)

FAIL: testChordWithoutKind (__main__.Test)
This verifies that the chord creation doesn't fail with an index out
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jwalls/music21/music21/harmony.py", line 3053, in testChordWithoutKind
    self.assertEqual('C3', str(cs.root()))
AssertionError: 'C3' != 'C4'
- C3
+ C4


FAIL: testPower (__main__.Test)
power chords should not have inversions
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jwalls/music21/music21/harmony.py", line 2991, in testPower
    self.assertEqual(pitches, cs1.pitches)
AssertionError: Tuples differ: (<music21.pitch.Pitch E2>, <music21.pitch.Pitch A2>, <music21.pitch.Pitch E3>) != (<music21.pitch.Pitch A2>, <music21.pitch.Pitch E3>)

First differing element 0:
<music21.pitch.Pitch E2>
<music21.pitch.Pitch A2>

First tuple contains 1 additional elements.
First extra element 2:
<music21.pitch.Pitch E3>

- (<music21.pitch.Pitch E2>, <music21.pitch.Pitch A2>, <music21.pitch.Pitch E3>)
?                       ^                         --------------------------

+ (<music21.pitch.Pitch A2>, <music21.pitch.Pitch E3>)
?                       ^


FAIL: testRootBassParsing (__main__.Test)
This tests a bug where the root and bass were wrongly parsed,
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jwalls/music21/music21/harmony.py", line 2827, in testRootBassParsing
    self.runTestOnChord(xmlString, figure, pitches)
  File "/Users/jwalls/music21/music21/harmony.py", line 2636, in runTestOnChord
    self.assertEqual(pitches, cs1.pitches)
AssertionError: Tuples differ: (<mus[14 chars]ch E-3>, <music21.pitch.Pitch E3>, <music21.pi[62 chars] D4>) != (<mus[14 chars]ch E-2>, <music21.pitch.Pitch E3>, <music21.pi[62 chars] D4>)

First differing element 0:
<music21.pitch.Pitch E-3>
<music21.pitch.Pitch E-2>

- (<music21.pitch.Pitch E-3>,
?                         ^

+ (<music21.pitch.Pitch E-2>,
?                         ^

   <music21.pitch.Pitch E3>,
   <music21.pitch.Pitch G#3>,
   <music21.pitch.Pitch B3>,
   <music21.pitch.Pitch D4>)

FAIL: testSus2Bass (__main__.Test)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jwalls/music21/music21/harmony.py", line 2955, in testSus2Bass
    self.assertEqual('E2', cs1.bass().nameWithOctave)
AssertionError: 'E2' != 'E'
- E2
?  -
+ E


FAIL: testWrongBass (__main__.Test)
Bug where bass isn't in chord.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jwalls/music21/music21/harmony.py", line 2921, in testWrongBass
    self.assertEqual(pitches, cs1.pitches)
AssertionError: Tuples differ: (<mus[13 chars]tch E-2>, <music21.pitch.Pitch C3>, <music21.p[36 chars] G3>) != (<mus[13 chars]tch E3>, <music21.pitch.Pitch G3>, <music21.pitch.Pitch C4>)

First differing element 0:
<music21.pitch.Pitch E-2>
<music21.pitch.Pitch E3>

First tuple contains 1 additional elements.
First extra element 3:
<music21.pitch.Pitch G3>

+ (<music21.pitch.Pitch E3>, <music21.pitch.Pitch G3>, <music21.pitch.Pitch C4>)
- (<music21.pitch.Pitch E-2>,
-  <music21.pitch.Pitch C3>,
-  <music21.pitch.Pitch E3>,
-  <music21.pitch.Pitch G3>)

----------------------------------------------------------------------
Ran 48 tests in 11.353s

FAILED (failures=8, errors=2)
</details>

- [ ] We should also merge the tests you contributed that already pass, plus any documentation improvements

I can help break this up into smaller patches to land as much of this as we can, and I'll of course add you as a co-author when using any of your material.

**Comment by mscuthbert** [comment]
Converted to draft/WIP -- feel free to revert when ready to review.

**Comment by mscuthbert** [comment]
Fixed in #792
